### PR TITLE
Fix reliability problems in KRB5 AS-REP format

### PR DESCRIPTION
**Pull request by kholia**
This fixes issue https://github.com/magnumripper/JohnTheRipper/issues/2821.

### CODE REVIEW

File: src/krb5_asrep_fmt_plug.c [hunk 1]
```diff
@@ -529,7 +529,8 @@ static int crypt_all(int *pcount, struct db_salt *salt)
 			}
 		}
 	}
-	new_keys = 0;
+	if (cur_salt->etype == 23)
+		new_keys = 0;
```
Comment (magnumripper) [inline src/krb5_asrep_fmt_plug.c]: I suggest instead move `new_keys = 0;` to line 448 or something

Comment (kholia) [reply src/krb5_asrep_fmt_plug.c]: Line 448 is `len = enc_to_utf16...` for the current code. I think you meant a different line number.

Comment (magnumripper) [reply src/krb5_asrep_fmt_plug.c]: I meant insert at line 448, moving the current line 448 down. It's pretty obvious if you look at the code there.

**Comment by kholia** [comment]
Here are two different patches,

``` diff
$ git diff
diff --git a/src/krb5_asrep_fmt_plug.c b/src/krb5_asrep_fmt_plug.c
index 61dc1c1..9a78813 100644
--- a/src/krb5_asrep_fmt_plug.c
+++ b/src/krb5_asrep_fmt_plug.c
@@ -477,6 +477,7 @@ static int crypt_all(int *pcount, struct db_salt *salt)
                                        any_cracked |= 1;
                                }
                        }
+                       new_keys = 0;
                }  else if (cur_salt->etype == 17 || cur_salt->etype == 18) {
                        // See "krb5int_decode_tgs_rep", "krb5int_enctypes_list", "krb5int_dk_decrypt" (key function),
                        // "krb5_k_decrypt", and "krb5_kdc_rep_decrypt_proc"
@@ -529,8 +530,6 @@ static int crypt_all(int *pcount, struct db_salt *salt)
                        }
                }
        }
-       if (cur_salt->etype == 23)
-               new_keys = 0;
 
        return count;
 }
```

This causes intermittent self-test failures. Setting `new_keys = 0` within the OpenMP block isn't safe nor correct.

``` diff
diff --git a/src/krb5_asrep_fmt_plug.c b/src/krb5_asrep_fmt_plug.c
index 61dc1c1..e9b04ae 100644
--- a/src/krb5_asrep_fmt_plug.c
+++ b/src/krb5_asrep_fmt_plug.c
@@ -445,6 +445,7 @@ static int crypt_all(int *pcount, struct db_salt *salt)
                                        UTF16 wkey[PLAINTEXT_LENGTH + 1];
                                        int len;
 
+                                       new_keys = 0;
                                        len = enc_to_utf16(wkey, PLAINTEXT_LENGTH,
                                                        (UTF8*)saved_key[index+i],
                                                        strlen(saved_key[index+i]));
@@ -529,8 +530,6 @@ static int crypt_all(int *pcount, struct db_salt *salt)
                        }
                }
        }
-       if (cur_salt->etype == 23)
-               new_keys = 0;
 
        return count;
 }
```

This patch moves `new_keys = 0` to line 448 but the format fails the self-tests always.

**Comment by magnumripper** [comment]
> Setting new_keys = 0 within the OpenMP block isn't safe nor correct.

Oh, right! Perhaps your original code is best/simplest.

**Comment by kholia** [comment]
Thanks!

**Pull request merged**